### PR TITLE
[SPARK-41442][SQL] Only update SQLMetric value if merging with valid metric

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch updates how `SQLMetric` merges two invalid instances where the value is both -1.

### Why are the changes needed?

We use -1 as initial value of `SQLMetric`, and change it to 0 while merging with other `SQLMetric` instances. A `SQLMetric` will be treated as invalid and filtered out later.

While we are developing with Spark, it is trouble behavior that two invalid `SQLMetric` instances merge to a valid `SQLMetric` because merging will set the value to 0.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @sunchao @dongjoon-hyun

**Review by sunchao — APPROVED** [review]
LGTM

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by viirya** [comment]
Thank you @sunchao @dongjoon-hyun !

Pending on CI (not sure if any existing tests depends on current behavior).

**Comment by dongjoon-hyun** [comment]
Could you re-trigger the failure CI test only?

**Comment by viirya** [comment]
I can reproduce the failed tests. Trying to fix them.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -2156,8 +2156,15 @@ class AdaptiveQueryExecSuite
         assert(aqeReads.length == 2)
         aqeReads.foreach { c =>
           val stats = c.child.asInstanceOf[QueryStageExec].getRuntimeStatistics
-          assert(stats.sizeInBytes >= 0)
+          val rowCount = stats.rowCount.get
           assert(stats.rowCount.get >= 0)
+          if (rowCount == 0) {
+            // For empty relation, the query stage doesn't serialize any bytes.
+            // The SQLMetric keeps initial value.
+            assert(stats.sizeInBytes == -1)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2164]: Nice!

**Review by dongjoon-hyun — APPROVED** [review]

**Review by HeartSaVioR — APPROVED** [review]
+1 It totally makes sense.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 2]
```diff
@@ -785,7 +785,11 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
 
     testMetricsInSparkPlanOperator(exchanges.head,
       Map("dataSize" -> 3200, "shuffleRecordsWritten" -> 100))
-    testMetricsInSparkPlanOperator(exchanges(1), Map("dataSize" -> 0, "shuffleRecordsWritten" -> 0))
+    // `testData2.filter($"b" === 0)` is an empty relation.
+    // The exchange doesn't serialize any bytes.
+    // The SQLMetric keeps initial value.
+    testMetricsInSparkPlanOperator(exchanges(1),
+      Map("dataSize" -> -1, "shuffleRecordsWritten" -> 0))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:792]: what do we show in the SQL UI? printing -1 as the size is a bit weird.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:792]: I think -1 will be filtered out.

**Pull request closed without merging**